src/lib-imap/imap-capability-list.h

```diff
@@ -0,0 +1,85 @@
+#ifndef IMAP_CAPABILITY_LIST_H
+#define IMAP_CAPABILITY_LIST_H
+
+#include <stdint.h>
+#include <stddef.h>
+
+#include "lib.h"
+#include "array.h"
+
+/* bitflags to represent different conditions of
+   the current imap session */
+enum imap_capability_visibility {
+	IMAP_CAP_VISIBILITY_NEVER           = 0,      /* never show this capability */
+	IMAP_CAP_VISIBILITY_PREAUTH         = BIT(0), /* show this cap before login */
+	IMAP_CAP_VISIBILITY_POSTAUTH        = BIT(1), /* show this cap after login */
+	IMAP_CAP_VISIBILITY_TLS_ACTIVE      = BIT(2), /* show this cap if tls is active */
+	IMAP_CAP_VISIBILITY_TLS_INACTIVE    = BIT(3), /* show this cap if tls is inactive */
+	IMAP_CAP_VISIBILITY_NO_LOGIN        = BIT(6), /* show this cap if login is disabled */
+	/* NOTE: connections from localhost are 'secure' regardless of SSL/TLS! */
+	IMAP_CAP_VISIBILITY_SECURE          = BIT(7), /* show this cap if we are secured */
+	IMAP_CAP_VISIBILITY_INSECURE        = BIT(8), /* show this cap if we aren't secured */
+
+	/* IMAP_CAP_VISIBILITY_REQUIRE_ALL indicates that the rest
+	   of the conditions must be met *exactly* to show this
+	   capability, otherwise if *at least* 1 condition is met
+	   the capability will be shown */
+	IMAP_CAP_VISIBILITY_REQUIRE_ALL     = BIT(31),
```

Comment:
Standard enums are signed ints. BIT(31) makes it unsigned, which isn't very standard anymore. I suppose should be BIT(30) instead. Also need to change IMAP_CAP_VISIBILITY_ALWAYS to 0x3FFFFFFF then.

Also, this would be better named with a "FLAG" in its name: IMAP_CAP_VISIBILITY_FLAG_REQUIRE_ALL

Reply:
Ah yes this was originally why I used the uint32_t in the lookup function (because of the sign bit), I will change it to use bit 30 instead.

I assume you also want to see "FLAG" in the other capability flags, I will add that as well.

Reply:
No, to me it seems like REQUIRE_ALL is special in a way from the other enums here? That's why I was thinking about the FLAG in it.

Reply:
Understood, I will add FLAG to the REQUIRE_ALL capability, and leave the rest.